corehq/apps/dump_reload/sql/load.py

```diff
@@ -37,39 +52,55 @@ class SqlDataLoader(DataLoader):
     slug = 'sql'
 
     def load_objects(self, object_strings, force=False):
-        # Keep a count of the installed objects
-        load_stats_by_db = {}
-        total_object_counts = []
-
-        def _process_chunk(chunk, total_object_counts=total_object_counts, load_stats_by_db=load_stats_by_db):
-            chunk_stats = load_objects(chunk)
-            total_object_counts.append(len(chunk))
-            self.stdout.write('Loaded {} SQL objects'.format(sum(total_object_counts)))
-            _update_stats(load_stats_by_db, chunk_stats)
-
-        chunk = []
-        for line in object_strings:
-            line = line.strip()
-            if not line:
-                continue
-            obj = json.loads(line)
-            if self.filter_object(obj):
-                chunk.append(obj)
-            if len(chunk) >= 1000:
-                _process_chunk(chunk)
-                chunk = []
-
-        if chunk:
-            _process_chunk(chunk)
-
-        _reset_sequences(list(load_stats_by_db.values()))
-
+        object_count = 0
+
+        def enqueue_object(dbalias_to_workerqueue, obj):
+            nonlocal object_count
+            db_alias = get_db_alias(obj)
+            __, queue = dbalias_to_workerqueue[db_alias]
+            queue.put(obj)
+            object_count += 1
+            if not object_count % 1000:
+                self.stdout.write(f'Loaded {object_count} SQL objects')
+
+        def terminate_workers(dbalias_to_workerqueue):
+            for __, queue in dbalias_to_workerqueue.values():
+                queue.put(None)
+
+        def collect_stats(dbalias_to_workerqueue):
+            worker_tasks = {wq[0] for wq in dbalias_to_workerqueue.values()}
+            return [t.result() for t in as_completed(worker_tasks)]
+
+        num_aliases = len(settings.DATABASES)
```

Comment:
Does this thing operate on all databases, or only shard dbs? If only shards, is this intentionally higher than the number of shards? Or maybe it doesn't matter?

Reply:
The original idea was to have a process for each shard, but the code already organises data by db_alias, so it was simpler to assign workers by db_alias instead. Capping the number of workers at "all of the databases" seemed safe. We probably don't have data for all the db_aliases, so we probably won't use all the workers.

Optimally we would want to have one process per core (which is the default value for `max_workers`), and if the number of db_aliases being loaded is greater than the number of cores, run them in different threads across those processes. ... But simple is better than complex.